source/tools/MatchBin.h

```diff
@@ -0,0 +1,241 @@
+
+  template<int numerator, int denominator>
+  struct ThreshSelector{
+    static emp::vector<size_t> Select(emp::vector<size_t>& uids, std::unordered_map<size_t, double>& scores, size_t n){
+        
+      double thresh = (double)numerator / (double)denominator;
+      unsigned int i = 0;
+      if (n < log2(uids.size())){
+        //Perform a bounded selection sort to find the first n results
+        while (i < n){ 
+          int minIndex = -1;
+          for(size_t j = i; j < uids.size(); ++j){
+            if (minIndex == -1 || scores.at(uids[j]) < scores.at(uids[minIndex])){
+              if (scores.at(uids[j]) < thresh){
+                minIndex = j;
+              }
+            }
+          }
+          if(minIndex == -1) break;
+          std::swap(uids.at(i),uids.at(minIndex));
+          ++i;
+        }
+      }
+      else{
+        std::sort(
+          uids.begin(),
+          uids.end(),
+          [&scores](size_t const &a, size_t const &b) {
+            return scores.at(a) < scores.at(b);
+          }
+        );
+        while(i < uids.size() && i < n && scores.at(uids[i]) < thresh) ++i;   
+      }
+      return emp::vector<size_t>(uids.begin(), uids.begin()+i);
+    }
+  };
+  
+  /// Factory function that generates a selector function that chooses
+  /// probabilisticly based on match quality with replacement.
+  struct RouletteSelector{
+    static emp::vector<size_t> Select(emp::vector<size_t>& uids, std::unordered_map<size_t, double>& scores, size_t n){
+      emp::Random random(1);
+      const double skew = 0.1;
+
+      IndexMap match_index(uids.size());
+      for (size_t i = 0; i < uids.size(); ++i) {
+        emp_assert(scores[uids[i]] >= 0);
+        match_index.Adjust(i, 1.0 / ( skew + scores[uids[i]] ));
+      }
+
+      emp::vector<size_t> res;
+      res.reserve(n);
+
+      for (size_t j = 0; j < n; ++j) {
+        const double match_pos = random.GetDouble(match_index.GetWeight());
+        const size_t idx = match_index.Index(match_pos);
+        res.push_back(uids[idx]);
+      }
+
+      return res;
+    }
+
+  };
+
+  /// A data container that allows lookup by tag similarity. It can be templated
+  /// on different tag types and is configurable on construction for (1) the
+  /// distance metric used to compute similarity between tags and (2) the
+  /// selector that is used to select which matches to return. Regulation
+  /// functionality is also provided, allowing dynamically adjustment of match
+  /// strength to a particular item (i.e., making all matches stronger/weaker).
+  /// A unique identifier is generated upon tag/item placement in the container.
+  /// This unique identifier can be used to view or edit the stored items and
+  /// their corresponding tags. Tag-based lookups return a list of matched
+  /// unique identifiers.
+  template <typename Val, typename Tag, typename Metric, typename Selector> class MatchBin {
+  private:
+
+    //const std::function<double(Tag, Tag)> metric;
```

Comment:
When we go through at the end to update the documentation and in-code comments, let's delete these.